text/0003-templates.md

```diff
@@ -0,0 +1,447 @@
+
+```
+node_modules/.pnpm/@example+patches/node_modules/@example/patches/patches/express@4.18.1.patch
+```
+
+### Combining Templates
+
+Templates may not be templated from other templates. Instead of an inheritance hierarchy, a package may refer to multiple templates with later entries in the list taking precedence.
+
+For example, if both `@organization/authoring-metadata` and `@team/authoring-metadata` have an `author` field, the value from `@team/authoring-metadata` will be used.
+
+```json5
+{
+  "name": "@example/simple",
+  "pnpm": {
+    "templates": {
+      "authoring": [
+        "@organization/authoring-metadata@0.1.0",
+        "@team/authoring-metadata@0.1.0"
+      ],
+    }
+  }
+}
+```
+
+Avoiding inheritance simplifies implementation and usability in several ways.
+
+- Suppose a new version of pnpm adds `pnpm.templates` config options. In an inheritance model, using these new options on a template would force consumers to a greater minimum version of pnpm.
+- Complex circular resolution during installation is avoided.
+- Template loading is more performant. The requirement to declare all required templates up front enables fetching in a single parallelized step, rather than a waterfall fetch at each level of a theoretical template inheritance hierarchy.
+- It becomes possible to reference multiple templates the package's author does not control with less ambiguity. In an inheritance-based mechanism, multiple inheritance would be necessary for such functionality. However, multiple inheritance can lead to [diamond-shaped problems](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem) and less clarity as to what the final value of the field should be.
+
+A few of the problems above can be mitigated by rendering the template package itself before publishing, but this introduces other problems. Toolkit templates install dependencies in the referencing package as `devDependencies`, which isn't desirable in other templates.
+
+The existing design is inspired by the concept of ["_composition over inheritance_"](https://en.wikipedia.org/wiki/Composition_over_inheritance), which provides reusability without defining difficult to change relationships between packages.
+
+### Specifiers
+
+Templates can be specified through:
+
+1. The `name@version` syntax to refer to an external template fetched from a registry.
+2. The `name@workspace:` syntax to refer to a template that's also a workspace package.
+3. The `file:<path>` syntax to refer to an on-disk path.
+
+### Shared Workspace Configuration
```

Comment:
This section actually took me a bit to think through. There's a decision point here with regard to whether we want to introduce "merging" of template configs.

Partitioning
============

A solution that would require no merging of configs requires users to create partitions of their workspace packages. The partition would be created with a config that looks like:

```json5
{
  "pnpm": {
    "workspaceTemplateConfigs": {
      // Applies to all packages in the workspace that's a "library"
      "library": {
        "authoring": [
          "@organization/authoring-metadata@0.1.0",
          "@team/authoring-metadata@0.1.0"
        ],
        "catalog": ["@example/frontend-catalog@workspace:"],
        "scripts": ["@example/library-scripts@workspace:"]
      },

      // Other packages may be tagged as an "application".
      "application": {
        "authoring": [
          "@organization/authoring-metadata@0.1.0",
          "@team/authoring-metadata@0.1.0"
        ],
        "catalog": ["@example/frontend-catalog@workspace:"],
        "scripts": ["@example/application-scripts@workspace:"]
      }

    },
  }
}
```

A package would only be able to reference one of the predefined template configs from the root `package.json`.

```json5
// packages/foo/package.json
{
  "name": "@example/foo",
  "pnpm": {
    "templates": {
      "workspaceConfig": "library"
    }
  }
}
```

Tagging
=======

A one-to-many relationship between template configs and packages may be too limiting. The duplication in the previous example is already showing. A many-to-many mechanism would allow less template config duplication.

```json5
// packages.json
{
  "pnpm": {
    "workspaceTemplateConfigs": {
      // All packages in the workspace should use this configuration.
      "default": {
        "authoring": [
          "@organization/authoring-metadata@0.1.0",
          "@team/authoring-metadata@0.1.0"
        ],
        "catalog": ["@example/frontend-catalog@workspace:"],
      },

      // Applies to all packages in the workspace tagged as a "library"
      "library": {
        "scripts": ["@example/library-scripts@workspace:"]
      },

      // Other packages may be tagged as an "application". For this example,
      // applications have a different set of scripts.
      "application": {
        "scripts": ["@example/application-scripts@workspace:"]
      }

    },
  }
}
```

```json5
// packages/foo/package.json
{
  "name": "@example/foo",
  "pnpm": {
    "templates": {
      "workspaceConfigs": ["default", "library"]
    }
  }
}
```

The downside is users would have to reason about how multiple template config blocks merge together. In the example of above, `default` and `library` cleanly merge, but that wouldn't be the case if both blocks specified an `authoring` flavor. Should the `authoring` arrays merge, or should the later configs completely replace all prior ones? This is unfortunate new complexity for users to have to reason about.

A `pnpm template view` command that shows the resulting `package.json` and how each field was determined would help, but users may still have to jump between multiple `package.json` files to reason about properties that were previously simple.

It's worth noting that from an implementation standpoint, merging configs is easy. The complexity to be mindful of is what's introduced to users.

Alternative Tagging Syntax
==========================

Instead of a tag → template mapping, an alternative design would be a template → tag mapping.

```json5
// packages.json
{
  "pnpm": {
    "workspaceTemplates": {
      "@organization/authoring-metadata@0.1.0": { "provides": "authoring", "tags": ["default"] },
      "@team/authoring-metadata@0.1.0": { "provides": "authoring", "tags": ["default"] },
      "@example/frontend-catalog@workspace:": { "provides": "catalog", "tags": ["default"] },

      "@example/library-scripts@workspace:": { "provides": "scripts", "tags": ["library"] },
      "@example/application-scripts@workspace:": { "provides": "scripts", "tags": ["library"] }
    },
  }
}
```

The main benefit is that template names don't need to be written more than once. The downside is that it's harder to tell which template applies to each package since users have to look at every entry and each `tags` field.

My thoughts are:

  - It probably is worthwhile to introduce some sort of template config merging.
  - The tag → template mapping is probably preferable over the reverse. In projects using a few amount of templates, template → tag looks nicer. But projects with many templates would be much harder to reason about.